File: Uncertainty_Estimation_Engine/metrics/UncertaintyMetrics.java
package com.uncertainty.metrics;

import java.util.HashMap;
import java.util.Map;

/**
 * Java implementation of uncertainty quality metrics.
 */
public class UncertaintyMetrics {

    public static double sharpness(double[] uncertainty) {
        double sum = 0;
        for (double u : uncertainty) sum += u;
        return sum / uncertainty.length;
    }

    public static double coverage(double[] y, double[] lower, double[] upper) {
        int covered = 0;
        for (int i = 0; i < y.length; i++) {
            if (y[i] >= lower[i] && y[i] <= upper[i]) covered++;
        }
        return (double) covered / y.length;
    }

    public static double meanIntervalWidth(double[] lower, double[] upper) {
        double total = 0;
        for (int i = 0; i < lower.length; i++) total += upper[i] - lower[i];
        return total / lower.length;
    }

    public static double gaussianNLL(double[] y, double[] mean, double[] std) {
        double nll = 0;
        for (int i = 0; i < y.length; i++) {
            double sigma2 = std[i] * std[i];
            nll += 0.5 * Math.log(2 * Math.PI * sigma2)
                 + Math.pow(y[i] - mean[i], 2) / (2 * sigma2);
        }
        return nll / y.length;
    }

    public static Map<String, Double> evaluateAll(
            double[] y, double[] mean, double[] std, double[] lower, double[] upper) {
        Map<String, Double> results = new HashMap<>();
        results.put("sharpness", sharpness(std));
        results.put("nll", gaussianNLL(y, mean, std));
        results.put("coverage", coverage(y, lower, upper));
        results.put("interval_width", meanIntervalWidth(lower, upper));
        return results;
    }
}

File: Uncertainty_Estimation_Engine/metrics/uncertainty_metrics.py
"""Comprehensive uncertainty quality metrics."""
import numpy as np
from typing import Dict


class UncertaintyMetrics:
    """Collects scalar quality measures for uncertainty estimates."""

    @staticmethod
    def sharpness(uncertainty: np.ndarray) -> float:
        """Mean uncertainty – lower is sharper."""
        return float(np.mean(uncertainty))

    @staticmethod
    def dispersion(uncertainty: np.ndarray) -> float:
        return float(np.std(uncertainty))

    @staticmethod
    def nll_gaussian(y: np.ndarray, mean: np.ndarray, std: np.ndarray) -> float:
        """Gaussian NLL – measures calibration of predictive distribution."""
        return float(np.mean(0.5 * np.log(2*np.pi*std**2) + (y-mean)**2 / (2*std**2)))

    @staticmethod
    def crps(y: np.ndarray, mean: np.ndarray, std: np.ndarray) -> float:
        """Continuous Ranked Probability Score (analytical Gaussian)."""
        from scipy.stats import norm
        z = (y - mean) / std
        return float(np.mean(std * (z*(2*norm.cdf(z)-1) + 2*norm.pdf(z) - 1/np.sqrt(np.pi))))

    @staticmethod
    def coverage(y: np.ndarray, lower: np.ndarray, upper: np.ndarray) -> float:
        return float(np.mean((y >= lower) & (y <= upper)))

    @staticmethod
    def interval_width(lower: np.ndarray, upper: np.ndarray) -> float:
        return float(np.mean(upper - lower))

    def evaluate_all(self, y, mean, std, lower=None, upper=None) -> Dict[str, float]:
        out = {
            "sharpness": self.sharpness(std),
            "dispersion": self.dispersion(std),
            "nll": self.nll_gaussian(y, mean, std),
            "crps": self.crps(y, mean, std),
        }
        if lower is not None and upper is not None:
            out["coverage"] = self.coverage(y, lower, upper)
            out["interval_width"] = self.interval_width(lower, upper)
        return out
